## Выполнил: Соковнин Игорь

# Спортивный анализ данных. Платформа Kaggle

## Урок 5. Feature Engineering, Feature Selection, part I

### Домашнее задание:

Продолжим работу с данными, которые были использованы в ДЗ2 и 3, продолжим решать задачу обнаружения мошеннических транзакций, что позволит получить полное решение задачи / полный пайплайн.

**[Задание 0](#task0):**
 Выбрать любую модель машинного обучения и зафиксировать любой тип валидации. Обучить базовую модель и зафиксировать базовое качество модели. В каждом следующем задании нужно будет обучить выбранную модель и оценивать ее качество на зафиксированной схеме валидации. После каждого задания, требуется сделать вывод о достигаемом качестве модели, по сравнению с качестом из предыдущего шага.

**[Задание 1](#task1):**
 Признак <font color='red'>TransactionDT</font> - это смещение в секундах относительно базовой даты. Базовая дата - <font color='red'>2017-12-01</font>, преобразовать признак <font color='red'>TransactionDT</font> в <font color='red'>datetime</font>, прибавив к базовой дате исходное значение признака. Из полученного признака выделить год, месяц, день недели, час, день.

**[Задание 2](#task2):**
 Сделать конкатенацию признаков
* <font color='red'>card1</font> + <font color='red'>card2</font>;
* <font color='red'>card1</font> + <font color='red'>card2</font> + <font color='red'>card_3</font> + <font color='red'>card_5</font>;
* <font color='red'>card1</font> + <font color='red'>card2</font> + <font color='red'>card_3</font> + <font color='red'>card_5</font> + <font color='red'>addr1</font> + <font color='red'>addr2</font>

Рассматривать их как категориальных признаки.

**[Задание 3](#task3):**
 Сделать <font color='red'>*FrequencyEncoder*</font> для признаков <font color='red'>*card1*</font> - <font color='red'>*card6*, *addr1*, *addr2*</font>.

**[Задание 4](#task4):**
 Создать признаки на основе отношения: <font color='red'>TransactionAmt</font> к вычисленной статистике. Статистика - среднее значение / стандартное отклонение <font color='red'>TransactionAmt</font>, сгруппированное по <font color='red'>card1</font> - <font color='red'>card6, addr1, addr2</font>, и по признакам, созданным в задании 2.

**[Задание 5](#task5):**
 Создать признаки на основе отношения: <font color='red'>D15</font> к вычисленной статистике. Статистика - среднее значение / стандартное отклонение <font color='red'>D15</font>, сгруппированное по <font color='red'>card1</font> - <font color='red'>card6, addr1, addr2</font>, и по признакам, созданным в задании 2.

**[Задание 6](#task6):**
 Выделить дробную часть и целую часть признака <font color='red'>TransactionAmt</font> в два отдельных признака. После создать отдельных признак - логарифм от <font color='red'>TransactionAmt</font>

**[Задание 7](#task7) (опция):**
 Выполнить предварительную подготовку / очистку признаков <font color='red'>P_emaildomain</font> и <font color='red'>R_emaildomain</font> (что и как делать - остается на ваше усмотрение) и сделать <font color='red'>Frequency Encoding</font> для очищенных признаков.

### Вывод по заданию:

- Относительно большое улучшение модели по сравнению с базовым решением дали созданные признаки из __Задания 2, 3, 7__.
- Улучшение модели по сравнению с базовым решением дали созданные признаки из __Задания 4, 5, 6__.
- Улучшение модели по сравнению с базовым решением не дали созданные признаки из __Задания 1__

Требуется дальнейший анализ.

Задание 0 (без обработки):
- bestTest = 0.8827161236
- bestIteration = 419


Задание 1:
- bestTest = 0.8812417137
- bestIteration = 455

Вывод:
- Добавление новых признаков (Задание 1) не дало улучшения качества модели по сравнению с базовым решением.


Задание 2:
- bestTest = 0.9216976237
- bestIteration = 557

Вывод:
- Добавление новых признаков (Задание 2) значительно улучшило качество модели по сравнению с базовым решением.

Задание 3:
- bestTest = 0.9180509792
- bestIteration = 506

Вывод:
- Добавление новых признаков (Задание 3) значительно улучшило качество модели по сравнению с базовым решением.


Задание 4:
- bestTest = 0.8842896115
- bestIteration = 442

Вывод:
- Добавление новых признаков (Задание 4) незначительно улучшило качества модели по сравнению с базовым решением.


Задание 5:
- bestTest = 0.8832494667
- bestIteration = 463

Вывод:
-  Добавление новых признаков (Задание 5) незначительно улучшило качества модели по сравнению с базовым решением.


Задание 6:
- bestTest = 0.8828945346
- bestIteration = 443

Вывод:
- Добавление новых признаков (Задание 6) незначительно улучшило качества модели по сравнению с базовым решением.


Задание 7:
- bestTest = 0.8859097396
- bestIteration = 458


Вывод:
- Добавление новых признаков (Задание 7) улучшило качества модели по сравнению с базовым решением.

## Подключение библиотек и скриптов

In [1]:
import datetime
import warnings
import numpy as np
import pandas as pd 
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)

# Модель
import xgboost as xgb
import catboost as cb

# Метрика
from sklearn.metrics import roc_auc_score, auc
from sklearn.model_selection import KFold, StratifiedKFold, train_test_split, cross_val_score

warnings.simplefilter("ignore")
%matplotlib inline

In [2]:
# разварачиваем выходной дисплей, чтобы увидеть больше столбцов и строк a pandas DataFrame
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)

In [3]:
def reduce_mem_usage(df):
    '''Сокращение размера датафрейма за счёт изменения типа данных'''
    
    start_mem = df.memory_usage().sum() / 1024**2
    print('Memory usage of dataframe is {:.2f} MB'.format(start_mem))
    
    for col in df.columns:
        col_type = df[col].dtype
        
        if col_type != object:
            c_min = df[col].min()
            c_max = df[col].max()
            if str(col_type)[:3] == 'int':
                if c_min > np.iinfo(np.int8).min and c_max < np.iinfo(np.int8).max:
                    df[col] = df[col].astype(np.int8)
                elif c_min > np.iinfo(np.int16).min and c_max < np.iinfo(np.int16).max:
                    df[col] = df[col].astype(np.int16)
                elif c_min > np.iinfo(np.int32).min and c_max < np.iinfo(np.int32).max:
                    df[col] = df[col].astype(np.int32)
                elif c_min > np.iinfo(np.int64).min and c_max < np.iinfo(np.int64).max:
                    df[col] = df[col].astype(np.int64)  
            else:
                if c_min > np.finfo(np.float32).min and c_max < np.finfo(np.float32).max:
                    df[col] = df[col].astype(np.float32)
                else:
                    df[col] = df[col].astype(np.float64)
        else:
            df[col] = df[col].astype('category')

    end_mem = df.memory_usage().sum() / 1024**2
    print('Memory usage after optimization is: {:.2f} MB'.format(end_mem))
    print('Decreased by {:.1f}%'.format(100 * (start_mem - end_mem) / start_mem))
    
    return df

In [4]:
!dir

 Том в устройстве C имеет метку Новый том
 Серийный номер тома: 6E3D-C99D

 Содержимое папки C:\Users\sil\Desktop\Python_for_DataSience\Спортивный анализ данных. Платформа Kaggle II\Урок 5. Feature Engineering, Feature Selection, part I\HW

29.03.2021  12:40    <DIR>          .
29.03.2021  12:40    <DIR>          ..
27.03.2021  14:05    <DIR>          .ipynb_checkpoints
28.03.2021  14:08    <DIR>          catboost_info
29.03.2021  02:02           239 022 lesson_5_hw - 2021-03-29.ipynb
29.03.2021  12:17           289 490 lesson_5_hw - 2021-03-29_1.ipynb
28.03.2021  16:49           163 768 lesson_5_hw 2021-03-28 CatBoost.ipynb
28.03.2021  13:39           118 506 lesson_5_hw 2021-03-28 XGBoost.ipynb
29.03.2021  12:38           289 544 lesson_5_hw.ipynb
29.03.2021  12:38         1 856 738 lesson_5_hw.pdf
29.03.2021  12:40           471 203 lesson_5_hw.rar
               7 файлов      3 428 271 байт
               4 папок  70 679 908 352 байт свободно


In [5]:
# input
TRAIN_DATASET_PATH = '../../data/assignment_2_train.csv'
TEST_DATASET_PATH = '../../data/assignment_2_test.csv'

## Загрузка данных

In [6]:
# Тренировочные данные
# train = pd.read_csv(TRAIN_DATASET_PATH, header = none)  # если надо скрыть названия столбцов
train = pd.read_csv(TRAIN_DATASET_PATH)
df_train =reduce_mem_usage(train)  # Уменьшаем размер данныхM
df_train.head(2)

Memory usage of dataframe is 541.08 MB
Memory usage after optimization is: 262.48 MB
Decreased by 51.5%


TransactionID  isFraud  TransactionDT  TransactionAmt ProductCD  card1  card2  card3       card4  card5   card6  addr1  addr2  dist1  dist2 P_emaildomain R_emaildomain   C1   C2   C3   C4   C5   C6   C7   C8   C9  C10  C11  C12  C13  C14    D1  D2    D3   D4  D5  D6  D7  D8  D9   D10   D11  D12  D13  D14  D15   M1   M2   M3  M4 M5 M6   M7   M8   M9   V1   V2   V3   V4   V5   V6   V7   V8   V9  V10  V11  V12  V13  V14  V15  V16  V17  V18  V19  V20  V21  V22  V23  V24  V25  V26  V27  V28  V29  V30  V31  V32  V33  V34  V35  V36  V37  V38  V39  V40  V41  V42  V43  V44  V45  V46  V47  V48  V49  V50  V51  V52  V53  V54  V55  V56  V57  V58  V59  V60  V61  V62  V63  V64  V65  V66  V67  V68  V69  V70  V71  V72  V73  V74  V75  V76  V77  V78  V79  V80  V81  V82  V83  V84  V85  V86  V87  V88  V89  V90  V91  V92  V93  V94  V95  V96  V97  V98  V99  V100  V101  V102  V103  V104  V105  V106  V107  V108  V109  V110  V111  V112  V113  V114  V115  V116  V117  V118  V119  V120  V121  V122  V123  \
0        2987000        0          86400            68.5         W  13926    NaN  150.0    discover  142.0  credit  315.0   87.0   19.0    NaN           NaN           NaN  1.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  2.0  0.0  1.0  1.0  14.0 NaN  13.0  NaN NaN NaN NaN NaN NaN  13.0  13.0  NaN  NaN  NaN  0.0    T    T    T  M2  F  T  NaN  NaN  NaN  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  1.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0   0.0   0.0   1.0   0.0   0.0   0.0   0.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
1        2987001        0          86401            29.0         W   2755  404.0  150.0  mastercard  102.0  credit  325.0   87.0    NaN    NaN     gmail.com           NaN  1.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  1.0   0.0 NaN   NaN  0.0 NaN NaN NaN NaN NaN   0.0   NaN  NaN  NaN  NaN  0.0  NaN  NaN  NaN  M0  T  T  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   

   V124  V125  V126   V127  V128  V129  V130  V131  V132   V133  V134  V135  V136  V137  V138  V139  V140  V141  V142  V143  V144  V145  V146  V147  V148  V149  V150  V151  V152  V153  V154  V155  V156  V157  V158  V159  V160  V161  V162  V163  V164  V165  V166  V167  V168  V169  V170  V171  V172  V173  V174  V175  V176  V177  V178  V179  V180  V181  V182  V183  V184  V185  V186  V187  V188  V189  V190  V191  V192  V193  V194  V195  V196  V197  V198  V199  V200  V201  V202  V203  V204  V205  V206  V207  V208  V209  V210  V211  V212  V213  V214  V215  V216  V217  V218  V219  V220  V221  V222  V223  V224  V225  V226  V227  V228  V229  V230  V231  V232  V233  V234  V235  V236  V237  V238  V239  V240  V241  V242  V243  V244  V245  V246  V247  V248  V249  V250  V251  V252  V253  V254  V255  V256  V257  V258  V259  V260  V261  V262  V263  V264  V265  V266  V267  V268  V269  V270  V271  V272  V273  V274  V275  V276  V277  V278  V279  V280  V281  V282  V283  V284  V285  V286  V287  V288  \
0   1.0   1.0   0

In [7]:
# Тестовые данные
leaderboard = pd.read_csv(TEST_DATASET_PATH)
df_test =reduce_mem_usage(leaderboard)  # Уменьшаем размер данных

df_test.head(2)

Memory usage of dataframe is 300.60 MB
Memory usage after optimization is: 145.83 MB
Decreased by 51.5%


TransactionID  isFraud  TransactionDT  TransactionAmt ProductCD  card1  card2  card3 card4  card5   card6  addr1  addr2  dist1  dist2 P_emaildomain R_emaildomain   C1   C2   C3   C4   C5   C6   C7   C8   C9  C10  C11  C12  C13  C14   D1   D2   D3   D4   D5  D6  D7  D8  D9  D10  D11  D12  D13  D14  D15   M1   M2   M3   M4   M5 M6   M7   M8   M9   V1   V2   V3   V4   V5   V6   V7   V8   V9  V10  V11  V12  V13  V14  V15  V16  V17  V18  V19  V20  V21  V22  V23  V24  V25  V26  V27  V28  V29  V30  V31  V32  V33  V34  V35  V36  V37  V38  V39  V40  V41  V42  V43  V44  V45  V46  V47  V48  V49  V50  V51  V52  V53  V54  V55  V56  V57  V58  V59  V60  V61  V62  V63  V64  V65  V66  V67  V68  V69  V70  V71  V72  V73  V74  V75  V76  V77  V78  V79  V80  V81  V82  V83  V84  V85  V86  V87  V88  V89  V90  V91  V92  V93  V94  V95  V96  V97  V98  V99  V100  V101  V102  V103  V104  V105  V106  V107  V108  V109  V110  V111  V112  V113  V114  V115  V116  V117  V118  V119  V120  V121  V122  V123  V124  \
0        3287000        1        7415038           226.0         W  12473  555.0  150.0  visa  226.0  credit  299.0   87.0  116.0    NaN       aol.com           NaN  2.0  3.0  0.0  0.0  0.0  5.0  0.0  0.0  3.0  0.0  3.0  2.0  6.0  2.0  4.0  4.0  0.0  4.0  3.0 NaN NaN NaN NaN  4.0  4.0  NaN  NaN  NaN  3.0    T    T    F   M0    T  F    F    F    T  1.0  2.0  2.0  1.0  3.0  1.0  1.0  1.0  2.0  0.0  0.0  2.0  2.0  1.0  0.0  0.0  0.0  0.0  2.0  5.0  0.0  0.0  1.0  1.0  1.0  2.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  2.0  2.0  0.0  0.0  1.0  0.0  0.0  1.0  1.0  1.0  2.0  0.0  0.0  0.0  1.0  1.0  2.0  2.0  2.0  2.0  0.0  0.0  0.0  0.0  2.0  5.0  0.0  0.0  1.0  1.0  2.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  0.0  1.0  1.0  1.0  0.0  0.0  0.0  1.0  3.0  0.0  0.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  2.0  7.0  7.0  0.0  1.0   1.0   2.0   6.0   6.0   0.0   0.0   0.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
1        3287001        0        7415054          3072.0         W  15651  417.0  150.0  visa  226.0   debit  330.0   87.0    NaN    NaN     yahoo.com           NaN  1.0  1.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  0.0  1.0  0.0  1.0  1.0  0.0  NaN  0.0  0.0  0.0 NaN NaN NaN NaN  0.0  NaN  NaN  NaN  NaN  0.0  NaN  NaN  NaN  NaN  NaN  T  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  1.0  1.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  1.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0  1.0  1.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  0.0  1.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  0.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   

   V125   V126    V127    V128  V129   V130   V131   V132    V133    V134  V135  V136  V137  V138  V139  V140  V141  V142  V143  V144  V145  V146  V147  V148  V149  V150  V151  V152  V153  V154  V155  V156  V157  V158  V159  V160  V161  V162  V163  V164  V165  V166  V167  V168  V169  V170  V171  V172  V173  V174  V175  V176  V177  V178  V179  V180  V181  V182  V183  V184  V185  V186  V187  V188  V189  V190  V191  V192  V193  V194  V195  V196  V197  V198  V199  V200  V201  V202  V203  V204  V205  V206  V207  V208  V209  V210  V211  V212  V213  V214  V215  V216  V217  V218  V219  V220  V221  V222  V223  V224  V225  V226  V227  V228  V229  V230  V231  V232  V233  V234  V235  V236  V237  V238  V239  V240  V241  V242  V243  V244  V245  V246  V247  V248  V249  V250  V251  V252  V253  V254  V255  V256  V257  V258  V259  V260  V261  V262  V263  V264  V265  V266  V267  V268  V269  V270  V271  V272  V273  V274  V275  V276  V277  V278  V279  V280  V281  V282  V283  V284  V285  V286  V287  V288  \
0   1.0  4

__Числовых признаки__

In [8]:
# Общее количество записей в датафрейме = 180 000
# Исключаем такие поля содержащие меньше 100 000 значений,
# из предполажения, что значение этих полей несущественно (всегда можно этот параметр проварьировать).
 
numerical_features = [
'TransactionID',  # Индекс
'isFraud',  # Целевой параметр
'TransactionDT',  # Временя совершения транзакции
'TransactionAmt',  # Сумма транзакции
'card1',
'card2',
'card3',
'card5',
'addr1',
'addr2',
'C1',
'C2',
'C3',
'C4',
'C5',
'C6',
'C7',
'C8',
'C9',
'C10',
'C11',
'C12',
'C13',
'C14',
'D1',
'D4',
'D10',
#'D11', ## <  50 000
'D15',
'V12',
'V13',
'V14',
'V15',
'V16',
'V17',
'V18',
'V19',
'V20',
'V21',
'V22',
'V23',
'V24',
'V25',
'V26',
'V27',
'V28',
'V29',
'V30',
'V31',
'V32',
'V33',
'V34',
'V35',
'V36',
'V37',
'V38',
'V39',
'V40',
'V41',
'V42',
'V43',
'V44',
'V45',
'V46',
'V47',
'V48',
'V49',
'V50',
'V51',
'V52',
'V53',
'V54',
'V55',
'V56',
'V57',
'V58',
'V59',
'V60',
'V61',
'V62',
'V63',
'V64',
'V65',
'V66',
'V67',
'V68',
'V69',
'V70',
'V71',
'V72',
'V73',
'V74',
'V75',
'V76',
'V77',
'V78',
'V79',
'V80',
'V81',
'V82',
'V83',
'V84',
'V85',
'V86',
'V87',
'V88',
'V89',
'V90',
'V91',
'V92',
'V93',
'V94',
'V95',
'V96',
'V97',
'V98',
'V99',
'V100',
'V101',
'V102',
'V103',
'V104',
'V105',
'V106',
'V107',
'V108',
'V109',
'V110',
'V111',
'V112',
'V113',
'V114',
'V115',
'V116',
'V117',
'V118',
'V119',
'V120',
'V121',
'V122',
'V123',
'V124',
'V125',
'V126',
'V127',
'V128',
'V129',
'V130',
'V131',
'V132',
'V133',
'V134',
'V135',
'V136',
'V137',
'V280',
'V281',
'V282',
'V283',
'V284',
'V285',
'V286',
'V287',
'V288',
'V289',
'V290',
'V291',
'V292',
'V293',
'V294',
'V295',
'V296',
'V297',
'V298',
'V299',
'V300',
'V301',
'V302',
'V303',
'V304',
'V305',
'V306',
'V307',
'V308',
'V309',
'V310',
'V311',
'V312',
'V313',
'V314',
'V315',
'V316',
'V317',
'V318',
'V319',
'V320',
'V321'
]

__Обрабатка категориальные признаков__

In [9]:
catigorical_features = [
'ProductCD',  # 180000 non-null  category
'card4',  # 179992 non-null  category
'card6',  # 179993 non-null  category
'P_emaildomain',  # 151560 non-null  category
'R_emaildomain',  # 60300 non-null   category
'M1',  # 61749 non-null   category
'M2',  # 61749 non-null   category
'M3',  # 61749 non-null   category
'M4',  # 83276 non-null   category
'M5',  # 61703 non-null   category
'M6',  # 105652 non-null  category
'M7',  # 31652 non-null   category
'M8',  # 31652 non-null   category
'M9'  # 31652 non-null   category
]

### Подготовка тренировочных данных

In [10]:
data = []
data = df_train[numerical_features + catigorical_features]

# заполняем пропуски в категориалиных признаках
for feature in catigorical_features:
    data[feature] = data[feature].cat.add_categories('Unknown')
    data[feature].fillna('Unknown', inplace =True) 
    
# Каждой категории сопоставляет целое число (номер категории) - https://dyakonov.org/2016/08/03/python-категориальные-признаки/
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
for cat_colname in data[catigorical_features].columns:
    le.fit(data[cat_colname])
    data[cat_colname+'_le'] = le.transform(data[cat_colname])

target = df_train["isFraud"]

In [11]:
df_train_new = data
#df_train_new = df_train_new.drop(catigorical_features, axis=1)
# df_train_new.columns

In [12]:
# df_train_new = df_train_new.drop(["TransactionID", "TransactionDT", "isFraud"], axis=1)

In [13]:
catigorical_features_new = ['ProductCD_le', 'card4_le', 'card6_le', 'R_emaildomain_le',
                      'M1_le', 'M2_le', 'M3_le', 'M4_le', 'M5_le', 'M6_le', 'M7_le', 'M8_le', 'M9_le']

### Подготовка тестовых данных

In [14]:
data = []
data = df_test[numerical_features + catigorical_features]

# заполняем пропуски в категориалиных признаках
for feature in catigorical_features:
    data[feature] = data[feature].cat.add_categories('Unknown')
    data[feature].fillna('Unknown', inplace =True) 
    
# Каждой категории сопоставляет целое число (номер категории) - https://dyakonov.org/2016/08/03/python-категориальные-признаки/
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
for cat_colname in data[catigorical_features].columns:
    le.fit(data[cat_colname])
    data[cat_colname+'_le'] = le.transform(data[cat_colname])

#target = df_train["isFraud"]

df_test_new = data
#f_test_new = df_test_new.drop(catigorical_features, axis=1)
df_test_new = df_test_new.drop(["TransactionID"], axis=1)

## Задание 0:<a class="anchor" id="task0"></a>

Выбрать любую модель машинного обучения и зафиксировать любой тип валидации. Обучить базовую модель и зафиксировать базовое качество модели. В каждом следующем задании нужно будет обучить выбранную модель и оценивать ее качество на зафиксированной схеме валидации. После каждого задания, требуется сделать вывод о достигаемом качестве модели, по сравнению с качестом из предыдущего шага.

__Hold-Out разбиение (Hold-Out валидация)__

In [15]:
data = df_train_new 
target = data["isFraud"]
#data = data.drop(["TransactionID", "TransactionDT", "isFraud"], axis=1)
data = data.drop(["TransactionID", "isFraud"], axis=1)

In [16]:
x_train, x_test = train_test_split(
    data, train_size=0.75, random_state=27
)
y_train, y_test = train_test_split(
    target, train_size=0.75, random_state=27
)
print("x_train.shape = {} rows, {} cols".format(*x_train.shape))
print("x_test.shape = {} rows, {} cols".format(*x_test.shape))

x_train.shape = 135000 rows, 222 cols
x_test.shape = 45000 rows, 222 cols


In [17]:
model = {}
train_scores = pd.DataFrame({"target": y_train})
test_scores = pd.DataFrame({"target": y_test})

In [18]:
#x_train.head(2)
#print(x_train.info())
#x_test.head(2)
#print(x_train.info())

### XGBoost на числовых признаках

In [19]:
xgb_numerical_features = numerical_features.copy()  # Создаём копию списка
xgb_numerical_features.remove('isFraud')
xgb_numerical_features.remove('TransactionID')
#xgb_numerical_features.remove('TransactionDT')

In [20]:
xgb_params = {
    "booster": "gbtree",
    "objective": "binary:logistic",
    "eval_metric": "auc",
    "n_estimators": 1000,
    "learning_rate": 0.1,
    "reg_lambda": 10,
    "max_depth": 4,
    "gamma": 10,
    "nthread": 6,
    "seed": 27
}

eval_sets= [
    (x_train[xgb_numerical_features], y_train),
    (x_test[xgb_numerical_features], y_test)
]

In [21]:
xgb_model = xgb.XGBClassifier(**xgb_params)

xgb_model.fit(
    y=y_train,
    X=x_train[xgb_numerical_features],
    early_stopping_rounds=50,
    eval_set=eval_sets,
    eval_metric="auc",
    verbose=10
)

model["XGBoost_gbtree_num_features"] = xgb_model

[0]	validation_0-auc:0.75709	validation_1-auc:0.74768
[10]	validation_0-auc:0.80798	validation_1-auc:0.79743
[20]	validation_0-auc:0.84054	validation_1-auc:0.82946
[30]	validation_0-auc:0.87095	validation_1-auc:0.86259
[40]	validation_0-auc:0.88017	validation_1-auc:0.87050
[50]	validation_0-auc:0.88913	validation_1-auc:0.87711
[60]	validation_0-auc:0.89620	validation_1-auc:0.88277
[70]	validation_0-auc:0.90007	validation_1-auc:0.88530
[80]	validation_0-auc:0.90428	validation_1-auc:0.88827
[90]	validation_0-auc:0.90599	validation_1-auc:0.88941
[100]	validation_0-auc:0.90792	validation_1-auc:0.89099
[110]	validation_0-auc:0.91035	validation_1-auc:0.89305
[120]	validation_0-auc:0.91163	validation_1-auc:0.89392
[130]	validation_0-auc:0.91163	validation_1-auc:0.89392
[140]	validation_0-auc:0.91163	validation_1-auc:0.89392
[150]	validation_0-auc:0.91163	validation_1-auc:0.89392
[160]	validation_0-auc:0.91163	validation_1-auc:0.89392
[166]	validation_0-auc:0.91163	validation_1-auc:0.89392


In [22]:
train_scores["XGBoost_gbtree_num_features"] = xgb_model.predict_proba(x_train[xgb_numerical_features])[:,1]
test_scores["XGBoost_gbtree_num_features"] = xgb_model.predict_proba(x_test[xgb_numerical_features])[:,1]

In [23]:
train_scores

target  XGBoost_gbtree_num_features
141582       0                     0.012777
131503       0                     0.013938
173925       0                     0.010473
177012       0                     0.002903
69958        0                     0.010226
...        ...                          ...
4848         0                     0.007321
14879        0                     0.007348
36680        0                     0.009374
118456       0                     0.003609
5139         0                     0.005172

[135000 rows x 2 columns]

### CatBoost на числовых признаках

In [24]:
import catboost as cb

In [25]:
cb_params = {
    "n_estimators": 1000,
    "loss_function": "Logloss",
    "eval_metric": "AUC",
    "task_type": "CPU",
    #"max_bin": 20,
    "verbose": 10,
    "max_depth": 6,
    "l2_leaf_reg": 100,
    "early_stopping_rounds": 50,
    "thread_count": 6,
    "random_seed": 42
}

eval_sets= [
    (x_train[xgb_numerical_features], y_train),
    (x_test[xgb_numerical_features], y_test)
]

In [26]:
cb_model = cb.CatBoostClassifier(**cb_params)
cb_model.fit(x_train[xgb_numerical_features], y_train, eval_set=eval_sets)

0:	test: 0.6536584	test1: 0.6509021	best: 0.6509021 (0)	total: 446ms	remaining: 7m 25s
10:	test: 0.7782015	test1: 0.7634376	best: 0.7687531 (7)	total: 1.49s	remaining: 2m 14s
20:	test: 0.8199294	test1: 0.8049216	best: 0.8049216 (20)	total: 2.71s	remaining: 2m 6s
30:	test: 0.8359524	test1: 0.8252769	best: 0.8252769 (30)	total: 3.58s	remaining: 1m 51s
40:	test: 0.8482519	test1: 0.8384418	best: 0.8384418 (40)	total: 4.54s	remaining: 1m 46s
50:	test: 0.8514080	test1: 0.8403066	best: 0.8403928 (47)	total: 5.4s	remaining: 1m 40s
60:	test: 0.8539646	test1: 0.8411689	best: 0.8411689 (60)	total: 6.05s	remaining: 1m 33s
70:	test: 0.8557345	test1: 0.8428050	best: 0.8431332 (69)	total: 6.75s	remaining: 1m 28s
80:	test: 0.8603778	test1: 0.8481003	best: 0.8481003 (80)	total: 7.44s	remaining: 1m 24s
90:	test: 0.8657723	test1: 0.8548091	best: 0.8548091 (90)	total: 8.05s	remaining: 1m 20s
100:	test: 0.8678208	test1: 0.8568615	best: 0.8568615 (100)	total: 8.58s	remaining: 1m 16s
110:	test: 0.8698583	tes

In [27]:
train_scores["CatBoost_num_features"] = cb_model.predict_proba(x_train[xgb_numerical_features])[:,1]
test_scores["CatBoost_num_features"] = cb_model.predict_proba(x_test[xgb_numerical_features])[:,1]

In [28]:
y_pred = cb_model.predict_proba(df_test_new[xgb_numerical_features])[:,1]

In [29]:
score = roc_auc_score(df_test_new["isFraud"], y_pred)
score

0.8513559235540431

## Задание 1:<a class="anchor" id="task1"></a>

Признак <font color='red'>TransactionDT</font> - это смещение в секундах относительно базовой даты. Базовая дата - <font color='red'>2017-12-01</font>, преобразовать признак <font color='red'>TransactionDT</font> в <font color='red'>datetime</font>, прибавив к базовой дате исходное значение признака. Из полученного признака выделить год, месяц, день недели, час, день.

### CatBoost на числовых признаках

In [30]:
import datetime

In [31]:
# Значение: datetime.datetime(2017, 4, 5, 0, 18, 51, 980187)
# now = datetime.datetime.now()
# base_date = datetime.datetime(2017, 10, 1)
# d = datetime.timedelta(seconds=11316)
# date = base_date + d
# print(now)
# print(date)
# print(date.year)
# print(date.month)
# print(date.day)
# print(date.hour)
# print(date.minute)
# print(date.second)
# print(date.weekday())

In [32]:
# def function(x):
#     return datetime.timedelta(seconds=x)

# df = pd.DataFrame({'TransactionDT': [86400, 86401, 86402]})
# df['DT'] = df['TransactionDT'].apply(function)
# df

In [33]:
def function(x):
    base_date = datetime.datetime(2017, 10, 1)
    new_date = base_date + datetime.timedelta(seconds=x)
    year = new_date.year
    month = new_date.month
    week_day = new_date.weekday()
    hour = new_date.hour
    day = new_date.day
    #return new_date, year, month, week_day, hour, day
    return year, month, week_day, hour, day

# df['new_date'], df['year'], df['month'], df['week_day'], df['hour'], df['day']  = zip(*df['TransactionDT'].map(function))
# df

In [34]:
x_train_task_1 = x_train[xgb_numerical_features + catigorical_features].copy()
x_test_task_1 = x_test[xgb_numerical_features + catigorical_features].copy()
#df_test_new_task_1 = df_test_new[['TransactionID', 'isFraud'] + xgb_numerical_features].copy()
df_test_new_task_1 = df_test_new[['isFraud'] + xgb_numerical_features + catigorical_features].copy()

# x_train_task_1['new_date'], 
x_train_task_1['year'], x_train_task_1['month'], x_train_task_1['week_day'], x_train_task_1['hour'], x_train_task_1['day']  = \
zip(*x_train_task_1['TransactionDT'].map(function))

# x_test_task_1['new_date'], 
x_test_task_1['year'], x_test_task_1['month'], x_test_task_1['week_day'], x_test_task_1['hour'], x_test_task_1['day']  = \
zip(*x_test_task_1['TransactionDT'].map(function))

In [35]:
df_test_new_task_1['year'], df_test_new_task_1['month'], df_test_new_task_1['week_day'], df_test_new_task_1['hour'], df_test_new_task_1['day']  = \
zip(*df_test_new_task_1['TransactionDT'].map(function))

#x_train_task_1.columns

In [36]:
task_1_fields = ['year', 'month', 'week_day', 'hour', 'day']

In [37]:
eval_sets= [
    (x_train_task_1[xgb_numerical_features + task_1_fields], y_train),
    (x_test_task_1[xgb_numerical_features + task_1_fields], y_test)
]

In [38]:
cb_model = cb.CatBoostClassifier(**cb_params)
cb_model.fit(x_train_task_1[xgb_numerical_features + task_1_fields], y_train, eval_set=eval_sets)

0:	test: 0.6667114	test1: 0.6618119	best: 0.6618119 (0)	total: 190ms	remaining: 3m 9s
10:	test: 0.7583015	test1: 0.7459275	best: 0.7459275 (10)	total: 1.47s	remaining: 2m 12s
20:	test: 0.8245631	test1: 0.8100912	best: 0.8100912 (20)	total: 2.43s	remaining: 1m 53s
30:	test: 0.8459280	test1: 0.8345029	best: 0.8345029 (30)	total: 3.21s	remaining: 1m 40s
40:	test: 0.8517896	test1: 0.8401610	best: 0.8401610 (40)	total: 3.81s	remaining: 1m 29s
50:	test: 0.8557151	test1: 0.8445617	best: 0.8445617 (50)	total: 4.67s	remaining: 1m 26s
60:	test: 0.8568242	test1: 0.8454821	best: 0.8459765 (57)	total: 5.5s	remaining: 1m 24s
70:	test: 0.8597014	test1: 0.8481595	best: 0.8481595 (70)	total: 6.18s	remaining: 1m 20s
80:	test: 0.8633124	test1: 0.8520402	best: 0.8521644 (79)	total: 6.86s	remaining: 1m 17s
90:	test: 0.8661622	test1: 0.8550718	best: 0.8550718 (90)	total: 7.57s	remaining: 1m 15s
100:	test: 0.8678117	test1: 0.8566115	best: 0.8566115 (100)	total: 8.28s	remaining: 1m 13s
110:	test: 0.8693647	te

In [ ]:
cb_model.fit(
    x_train_task_1[xgb_numerical_features + task_1_fields],
    y_train,
    cat_features = xgb_numerical_features + task_1_fields,
    eval_set=eval_sets)

Задание 0:
- bestTest = 0.8827161236
- bestIteration = 419

Задание 1:
- bestTest = 0.8812417137
- bestIteration = 455

Вывод:
- Добавление новых признаков (Задание 1) не дало улучшения качества модели.

In [39]:
train_scores["CatBoost_task1_features"] = cb_model.predict_proba(x_train_task_1[xgb_numerical_features + task_1_fields])[:,1]
test_scores["CatBoost_task1_features"] = cb_model.predict_proba(x_test_task_1[xgb_numerical_features + task_1_fields])[:,1]
#train_scores["CatBoost_num_features"] = cb_model.predict_proba(x_train_task_1[xgb_numerical_features + task_1_fields])[:,1]
#test_scores["CatBoost_num_features"] = cb_model.predict_proba(x_test_task_1[xgb_numerical_features + task_1_fields])[:,1]

In [40]:
y_pred = cb_model.predict_proba(df_test_new_task_1[['isFraud'] + xgb_numerical_features + task_1_fields])[:,1]

In [41]:
score = roc_auc_score(df_test_new_task_1["isFraud"], y_pred)
score

0.8541448109129632

Задание 0:
- 0.8513559235540431

Задание 1:
- 0.8541448109129632

Вывод:
- Добавление новых признаков улучшило качество модели.


## Задание 2:<a class="anchor" id="task2"></a>

Cделать конкатенацию признаков
* <font color='red'>card1</font> + <font color='red'>card2</font>;
* <font color='red'>card1</font> + <font color='red'>card2</font> + <font color='red'>card_3</font> + <font color='red'>card_5</font>;
* <font color='red'>card1</font> + <font color='red'>card2</font> + <font color='red'>card_3</font> + <font color='red'>card_5</font> + <font color='red'>addr1</font> + <font color='red'>addr2</font>

Рассматривать их как категориальных признаки.

In [42]:
# import pandas as pd
# df = pd.DataFrame({'foo':['a','b','c'], 'bar':[1, 2, 3]})
# df['baz'] = df.agg(lambda x: f"{x['bar']} is {x['foo']}", axis=1)
# df

In [43]:
x_train_task_1.columns

Index(['TransactionDT', 'TransactionAmt', 'card1', 'card2', 'card3', 'card5', 'addr1', 'addr2', 'C1', 'C2',
       ...
       'M5', 'M6', 'M7', 'M8', 'M9', 'year', 'month', 'week_day', 'hour', 'day'], dtype='object', length=213)

In [44]:
# x_train_task_1['card1_card2'] = x_train_task_1.agg(lambda x: x['card1'] + x['card2'], axis=1)
# x_train_task_1['card1_card2_card_3_card_5'] = \
#         x_train_task_1.agg(lambda x: x['card1_card2'] + x['card5'] + x['card5'], axis=1)
# x_train_task_1['card1_card2_card_3_card_5_addr1_addr2'] = \
#             x_train_task_1.agg(lambda x: f"{x['card1_card2_card_3_card_5']} + {x['addr1']} + {x['addr2']}", axis=1)

x_train_task_1['card1_card2'] = x_train_task_1.agg(lambda x: f"{x['card1']} {x['card2']}", axis=1)
x_train_task_1['card1_card2_card_3_card_5'] = \
    x_train_task_1.agg(lambda x: f"{x['card1_card2']} {x['card3']} {x['card5']}", axis=1)
x_train_task_1['card1_card2_card_3_card_5_addr1_addr2'] = \
    x_train_task_1.agg(lambda x: f"{x['card1_card2_card_3_card_5']} {x['addr1']} {x['addr2']}", axis=1)

# x_train_task_1[['card1_card2', 'card1_card2_card_3_card_5', 'card1_card2_card_3_card_5_addr1_addr2']].head(2)

In [45]:
x_train_task_1[['card1_card2', 'card1_card2_card_3_card_5', 'card1_card2_card_3_card_5_addr1_addr2', 'year', 'hour', 'day', 'week_day', 'month']].head(2)

card1_card2 card1_card2_card_3_card_5 card1_card2_card_3_card_5_addr1_addr2  year  hour  day  week_day  month
141582  6892 560.0    6892 560.0 150.0 226.0     6892 560.0 150.0 226.0 433.0 87.0  2017    18    3         4     11
131503  2922 583.0    2922 583.0 150.0 226.0     2922 583.0 150.0 226.0 299.0 87.0  2017     2   31         1     10

In [46]:
x_test_task_1['card1_card2'] = x_test_task_1.agg(lambda x: f"{x['card1']} {x['card2']}", axis=1)
x_test_task_1['card1_card2_card_3_card_5'] = \
    x_test_task_1.agg(lambda x: f"{x['card1_card2']} {x['card3']} {x['card5']}", axis=1)
x_test_task_1['card1_card2_card_3_card_5_addr1_addr2'] = \
    x_test_task_1.agg(lambda x: f"{x['card1_card2_card_3_card_5']} {x['addr1']} {x['addr2']}", axis=1)

In [47]:
x_test_task_1[['card1_card2', 'card1_card2_card_3_card_5', 'card1_card2_card_3_card_5_addr1_addr2', 'year', 'hour', 'day', 'week_day', 'month']].head(2)

card1_card2 card1_card2_card_3_card_5 card1_card2_card_3_card_5_addr1_addr2  year  hour  day  week_day  month
78715  15186 480.0   15186 480.0 150.0 224.0    15186 480.0 150.0 224.0 299.0 87.0  2017    19   20         4     10
907     6019 583.0    6019 583.0 150.0 226.0     6019 583.0 150.0 226.0 225.0 87.0  2017     6    2         0     10

In [48]:
x_test_task_1['card1_card2'] = x_test_task_1.agg(lambda x: f"{x['card1']} {x['card2']}", axis=1)
x_test_task_1['card1_card2_card_3_card_5'] = \
    x_test_task_1.agg(lambda x: f"{x['card1_card2']} {x['card3']} {x['card5']}", axis=1)
x_test_task_1['card1_card2_card_3_card_5_addr1_addr2'] = \
    x_test_task_1.agg(lambda x: f"{x['card1_card2_card_3_card_5']} {x['addr1']} {x['addr2']}", axis=1)

In [49]:
# x_train_task_1.info()
categorical_features = x_train_task_1.select_dtypes(include=["object"]).columns
x_train_task_1[categorical_features] = x_train_task_1[categorical_features].astype(str)
x_test_task_1[categorical_features] = x_test_task_1[categorical_features].astype(str)
#categorical_features = []
categorical_features = ['card1_card2', 'card1_card2_card_3_card_5', 'card1_card2_card_3_card_5_addr1_addr2']
categorical_features

['card1_card2',
 'card1_card2_card_3_card_5',
 'card1_card2_card_3_card_5_addr1_addr2']

In [50]:
# x_test_task_1 = x_test[xgb_numerical_features].copy()
# df_test_new_task_1 = df_test_new[['TransactionID', 'isFraud'] + xgb_numerical_features].copy()
# df_test_new_task_1 = df_test_new[['isFraud'] + xgb_numerical_features].copy()

### CatBoost с категориальными признаками

In [51]:
# eval_sets= [
#     (x_train_task_1[xgb_numerical_features + task_1_fields + categorical_features], y_train),
#     (x_test_task_1[xgb_numerical_features + task_1_fields + categorical_features], y_test)
# ]
eval_sets= [
    (x_train_task_1[xgb_numerical_features + categorical_features], y_train),
    (x_test_task_1[xgb_numerical_features + categorical_features], y_test)
]

In [52]:
# cb_model.fit(
#     x_train_task_1[xgb_numerical_features + task_1_fields + categorical_features],
#     y_train,
#     cat_features = categorical_features,
#     eval_set=eval_sets)
cb_model.fit(
    x_train_task_1[xgb_numerical_features + categorical_features],
    y_train,
    cat_features = categorical_features,
    eval_set=eval_sets)

0:	test: 0.6169405	test1: 0.6013935	best: 0.6013935 (0)	total: 530ms	remaining: 8m 49s
10:	test: 0.7872496	test1: 0.7697118	best: 0.7697118 (10)	total: 3.65s	remaining: 5m 28s
20:	test: 0.8188597	test1: 0.8034291	best: 0.8034291 (20)	total: 5.4s	remaining: 4m 11s
30:	test: 0.8414185	test1: 0.8284577	best: 0.8284577 (30)	total: 9.29s	remaining: 4m 50s
40:	test: 0.8625035	test1: 0.8473969	best: 0.8478863 (39)	total: 12.9s	remaining: 5m
50:	test: 0.9180574	test1: 0.8809312	best: 0.8809312 (50)	total: 15.9s	remaining: 4m 55s
60:	test: 0.9245797	test1: 0.8846792	best: 0.8846792 (60)	total: 18.8s	remaining: 4m 49s
70:	test: 0.9267131	test1: 0.8856667	best: 0.8857395 (68)	total: 21.2s	remaining: 4m 37s
80:	test: 0.9278434	test1: 0.8857564	best: 0.8859937 (76)	total: 23.6s	remaining: 4m 27s
90:	test: 0.9290020	test1: 0.8884479	best: 0.8884479 (90)	total: 26.7s	remaining: 4m 26s
100:	test: 0.9299823	test1: 0.8911254	best: 0.8911254 (100)	total: 29.9s	remaining: 4m 26s
110:	test: 0.9317724	test1

Задание 0:
- bestTest = 0.8827161236
- bestIteration = 419

Задание 2:
- bestTest = 0.9205033376
- bestIteration = 497

Вывод:
- Добавление новых признаков (Задание 2) значительно улучшило качество модели по сравнению с базовым решением.


In [53]:
# train_scores["CatBoost_task2_features"] = \
#     cb_model.predict_proba(x_train_task_1[xgb_numerical_features + task_1_fields + categorical_features])[:,1]
train_scores["CatBoost_task2_features"] = \
    cb_model.predict_proba(x_train_task_1[xgb_numerical_features + categorical_features])[:,1]

In [54]:
# test_scores["CatBoost_task2_features"] = \
#     cb_model.predict_proba(x_test_task_1[xgb_numerical_features + task_1_fields + categorical_features])[:,1]
test_scores["CatBoost_task2_features"] = \
    cb_model.predict_proba(x_test_task_1[xgb_numerical_features + categorical_features])[:,1]

## Задание 3:<a class="anchor" id="task3"></a>

Сделать <font color='red'>*Frequency Encoding*</font> для признаков <font color='red'>*card1*</font> - <font color='red'>*card6*, *addr1*, *addr2*</font>.

См. "Урок 4 Предварительная обработка признаков/Категориальные признаки/Второй способ". Файл webinar4_features_part1.ipynb.

In [55]:
data = []
data_test = []
data = x_train_task_1.copy()
data_test = x_test_task_1.copy()

In [56]:
freq_encoder = data["card1"].value_counts(normalize=True)
data["card1_freq_enc"] = data["card1"].map(freq_encoder)
freq_encoder = data["card2"].value_counts(normalize=True)
data["card2_freq_enc"] = data["card2"].map(freq_encoder)
freq_encoder = data["card3"].value_counts(normalize=True)
data["card3_freq_enc"] = data["card3"].map(freq_encoder)
freq_encoder = data["card4"].value_counts(normalize=True)
data["card4_freq_enc"] = data["card4"].map(freq_encoder)
freq_encoder = data["card5"].value_counts(normalize=True)
data["card5_freq_enc"] = data["card5"].map(freq_encoder)
freq_encoder = data["card6"].value_counts(normalize=True)
data["card6_freq_enc"] = data["card6"].map(freq_encoder)
freq_encoder = data["addr1"].value_counts(normalize=True)
data["addr1_freq_enc"] = data["addr1"].map(freq_encoder)
freq_encoder = data["addr2"].value_counts(normalize=True)
data["addr2_freq_enc"] = data["addr2"].map(freq_encoder)
# https://towardsdatascience.com/all-about-categorical-variable-encoding-305f3361fd02
# fe = data.groupby('card1').size()/len(data)   
# data.loc[:, 'card1_freq_enc'] = data['card1'].map(fe)
# fe = data.groupby('card2').size()/len(data)   
# data.loc[:, 'card2_freq_enc'] = data['card2'].map(fe)
# fe = data.groupby('card3').size()/len(data)   
# data.loc[:, 'card3_freq_enc'] = data['card3'].map(fe)
# fe = data.groupby('card4').size()/len(data)   
# data.loc[:, 'card4_freq_enc'] = data['card4'].map(fe)
# fe = data.groupby('card5').size()/len(data)   
# data.loc[:, 'card5_freq_enc'] = data['card5'].map(fe)
# fe = data.groupby('card6').size()/len(data)   
# data.loc[:, 'card6_freq_enc'] = data['card6'].map(fe)
# fe = data.groupby('addr1').size()/len(data)   
# data.loc[:, 'addr1_freq_enc'] = data['addr1'].map(fe)
# fe = data.groupby('addr2').size()/len(data)   
# data.loc[:, 'addr2_freq_enc'] = data['addr2'].map(fe)

In [57]:
data[['card1', 'card1_freq_enc', 'card2', 'card2_freq_enc', 'card3', 'card3_freq_enc', \
     'card4', 'card4_freq_enc', 'card5', 'card5_freq_enc', 'card6', 'card6_freq_enc', \
     'addr1', 'addr1_freq_enc', 'addr2', 'addr2_freq_enc']].head(2)
# Функция map применяет функцию к каждому элементу последовательности и возвращает итератор с результатами.

card1  card1_freq_enc  card2  card2_freq_enc  card3  card3_freq_enc card4 card4_freq_enc  card5  card5_freq_enc   card6  card6_freq_enc  addr1  addr1_freq_enc  addr2  addr2_freq_enc
141582   6892        0.000311  560.0        0.000436  150.0        0.879139  visa       0.658237  226.0         0.51426  credit        0.317059  433.0        0.013992   87.0        0.982378
131503   2922        0.000104  583.0        0.054646  150.0        0.879139  visa       0.658237  226.0         0.51426  credit        0.317059  299.0        0.096911   87.0        0.982378

In [58]:
# https://towardsdatascience.com/all-about-categorical-variable-encoding-305f3361fd02
# fe = data_test.groupby('card1').size()/len(data_test)   
# data_test.loc[:, 'card1_freq_encode'] = data_test['card1'].map(fe)
# fe = data_test.groupby('card2').size()/len(data_test)   
# data_test.loc[:, 'card2_freq_encode'] = data_test['card2'].map(fe)
# fe = data_test.groupby('card3').size()/len(data_test)   
# data_test.loc[:, 'card3_freq_encode'] = data_test['card3'].map(fe)
# fe = data_test.groupby('card4').size()/len(data_test)   
# data_test.loc[:, 'card4_freq_encode'] = data_test['card4'].map(fe)
# fe = data_test.groupby('card5').size()/len(data_test)   
# data_test.loc[:, 'card5_freq_encode'] = data_test['card5'].map(fe)
# fe = data_test.groupby('card6').size()/len(data_test)   
# data_test.loc[:, 'card6_freq_encode'] = data_test['card6'].map(fe)
# fe = data_test.groupby('addr1').size()/len(data_test)   
# data_test.loc[:, 'addr1_freq_encode'] = data_test['addr1'].map(fe)
# fe = data_test.groupby('addr2').size()/len(data_test)   
# data_test.loc[:, 'addr2_freq_encode'] = data_test['addr2'].map(fe)
freq_encoder = data_test["card1"].value_counts(normalize=True)
data_test["card1_freq_enc"] = data_test["card1"].map(freq_encoder)
freq_encoder = data_test["card2"].value_counts(normalize=True)
data_test["card2_freq_enc"] = data_test["card2"].map(freq_encoder)
freq_encoder = data_test["card3"].value_counts(normalize=True)
data_test["card3_freq_enc"] = data_test["card3"].map(freq_encoder)
freq_encoder = data_test["card4"].value_counts(normalize=True)
data_test["card4_freq_enc"] = data_test["card4"].map(freq_encoder)
freq_encoder = data_test["card5"].value_counts(normalize=True)
data_test["card5_freq_enc"] = data_test["card5"].map(freq_encoder)
freq_encoder = data_test["card6"].value_counts(normalize=True)
data_test["card6_freq_enc"] = data_test["card6"].map(freq_encoder)
freq_encoder = data_test["addr1"].value_counts(normalize=True)
data_test["addr1_freq_enc"] = data_test["addr1"].map(freq_encoder)
freq_encoder = data_test["addr2"].value_counts(normalize=True)
data_test["addr2_freq_enc"] = data_test["addr2"].map(freq_encoder)

In [59]:
data_test[['card1', 'card1_freq_enc', 'card2', 'card2_freq_enc', 'card3', 'card3_freq_enc', \
     'card4', 'card4_freq_enc', 'card5', 'card5_freq_enc', 'card6', 'card6_freq_enc', \
     'addr1', 'addr1_freq_enc', 'addr2', 'addr2_freq_enc']].head(2)
# Функция map применяет функцию к каждому элементу последовательности и возвращает итератор с результатами.

card1  card1_freq_enc  card2  card2_freq_enc  card3  card3_freq_enc       card4 card4_freq_enc  card5  card5_freq_enc   card6  card6_freq_enc  addr1  addr1_freq_enc  addr2  addr2_freq_enc
78715  15186        0.000267  480.0        0.003451  150.0        0.881531  mastercard       0.303644  224.0        0.128824   debit        0.679267  299.0        0.099828   87.0        0.982243
907     6019        0.018267  583.0        0.055197  150.0        0.881531        visa       0.654067  226.0        0.515966  credit        0.320578  225.0        0.010197   87.0        0.982243

In [118]:
# task3_cat_features = ['card1_freq_encode', 'card2_freq_encode', 'card3_freq_encode', \
#      'card4_freq_encode', 'card5_freq_encode', 'card6_freq_encode', 'addr1_freq_encode', 'addr2_freq_encode']
# categorical_features = categorical_features + task3_cat_features
categorical_features = ['card1_card2',
 'card1_card2_card_3_card_5',
 'card1_card2_card_3_card_5_addr1_addr2',
 'card1_freq_enc',
 'card2_freq_enc',
 'card3_freq_enc',
 'card4_freq_enc',
 'card5_freq_enc',
 'card6_freq_enc',
 'addr1_freq_enc',
 'addr2_freq_enc',
 'card4',
 'card6'
]
#categorical_features = x_train_task_3.select_dtypes(include=["object"]).columns

In [61]:
x_train_task_3 = data[xgb_numerical_features + task_1_fields + categorical_features].copy()

In [62]:
x_train_task_3["card4"].head(2)

141582    visa
131503    visa
Name: card4, dtype: category
Categories (5, object): ['american express', 'discover', 'mastercard', 'visa', 'Unknown']

In [63]:
x_train_task_3[categorical_features] = x_train_task_3[categorical_features].astype(str)

In [64]:
x_test_task_3 = data_test[xgb_numerical_features + task_1_fields + categorical_features].copy()
x_test_task_3[categorical_features] = x_test_task_3[categorical_features].astype(str)

In [65]:
#x_test_task_3.isnull().sum(axis = 0)

In [119]:
# eval_sets= [
#     (x_train_task_3[xgb_numerical_features + task_1_fields + categorical_features], y_train),
#     (x_test_task_3[xgb_numerical_features + task_1_fields + categorical_features], y_test)
# ]
eval_sets= [
    (x_train_task_3[xgb_numerical_features + categorical_features], y_train),
    (x_test_task_3[xgb_numerical_features + categorical_features], y_test)
]

In [120]:
# cb_model.fit(
#     x_train_task_3[xgb_numerical_features + task_1_fields + categorical_features],
#     y_train,
#     cat_features = categorical_features,
#     eval_set=eval_sets)

cb_model.fit(
    x_train_task_3[xgb_numerical_features + categorical_features],
    y_train,
    cat_features = categorical_features,
    eval_set=eval_sets)

0:	test: 0.6495082	test1: 0.4114233	best: 0.4114233 (0)	total: 745ms	remaining: 12m 24s
10:	test: 0.7880146	test1: 0.7277898	best: 0.7584509 (8)	total: 4.25s	remaining: 6m 22s
20:	test: 0.8285493	test1: 0.8307867	best: 0.8307867 (20)	total: 7.05s	remaining: 5m 28s
30:	test: 0.8470180	test1: 0.8354330	best: 0.8380251 (28)	total: 10.2s	remaining: 5m 18s
40:	test: 0.8576627	test1: 0.8261190	best: 0.8380251 (28)	total: 13.5s	remaining: 5m 15s
50:	test: 0.8862479	test1: 0.8623361	best: 0.8623361 (50)	total: 16.3s	remaining: 5m 3s
60:	test: 0.9139052	test1: 0.8819011	best: 0.8819011 (60)	total: 19.3s	remaining: 4m 56s
70:	test: 0.9222446	test1: 0.8862534	best: 0.8862534 (70)	total: 22.1s	remaining: 4m 49s
80:	test: 0.9252371	test1: 0.8875602	best: 0.8875602 (80)	total: 25s	remaining: 4m 44s
90:	test: 0.9272170	test1: 0.8892153	best: 0.8893226 (88)	total: 27.9s	remaining: 4m 39s
100:	test: 0.9287967	test1: 0.8920581	best: 0.8920581 (100)	total: 31.1s	remaining: 4m 36s
110:	test: 0.9298694	tes

Задание 0:
- bestTest = 0.8827161236
- bestIteration = 419

Задание 3:
- bestTest = 0.9180509792
- bestIteration = 506


Вывод:
- Добавление новых признаков (Задание 3) значительно улучшило качество модели по сравнению с базовым решением.

## Задание 4:<a class="anchor" id="task4"></a>

Создать признаки на основе отношения: <font color='red'>TransactionAmt</font> к вычисленной статистике. Статистика - среднее значение / стандартное отклонение <font color='red'>TransactionAmt</font>, сгруппированное по <font color='red'>card1</font> - <font color='red'>card6, addr1, addr2</font>, и по признакам, созданным в задании 2.


In [68]:
# Leveraging Machine Learning to Detect Fraud: Tips to Developing a Winning Kaggle Solution
# https://developer.nvidia.com/blog/leveraging-machine-learning-to-detect-fraud-tips-to-developing-a-winning-kaggle-solution/
# temp = df.groupby('card1')['TransactionAmt'].agg(['mean']).rename({'mean':'TransactionAmt_card1_mean'},axis=1)
# df = pd.merge(df,temp,on='card1',how='left')

In [121]:
x_train_task_4 = []
x_test_task_4 = []
x_train_task_4 = x_train_task_3.copy()
x_test_task_4 = x_test_task_3.copy()

In [122]:
temp = x_train_task_4.groupby('card1')['TransactionAmt'].agg(['mean']).rename({'mean':'TransactionAmt_card1_mean'},axis=1)
x_train_task_4 = pd.merge(x_train_task_4,temp,on='card1',how='left')
temp = x_train_task_4.groupby('card2')['TransactionAmt'].agg(['mean']).rename({'mean':'TransactionAmt_card2_mean'},axis=1)
x_train_task_4 = pd.merge(x_train_task_4,temp,on='card2',how='left')
temp = x_train_task_4.groupby('card3')['TransactionAmt'].agg(['mean']).rename({'mean':'TransactionAmt_card3_mean'},axis=1)
x_train_task_4 = pd.merge(x_train_task_4,temp,on='card3',how='left')
temp = x_train_task_4.groupby('card5')['TransactionAmt'].agg(['mean']).rename({'mean':'TransactionAmt_card5_mean'},axis=1)
x_train_task_4 = pd.merge(x_train_task_4,temp,on='card5',how='left')
temp = x_train_task_4.groupby('card4')['TransactionAmt'].agg(['mean']).rename({'mean':'TransactionAmt_card4_mean'},axis=1)
x_train_task_4 = pd.merge(x_train_task_4,temp,on='card4',how='left')
temp = x_train_task_4.groupby('card6')['TransactionAmt'].agg(['mean']).rename({'mean':'TransactionAmt_card6_mean'},axis=1)
x_train_task_4 = pd.merge(x_train_task_4,temp,on='card6',how='left')

In [123]:
temp = x_train_task_4.groupby('card1_card2')['TransactionAmt'].agg(['mean']).\
rename({'mean':'TransactionAmt_card1_card2_mean'},axis=1)
x_train_task_4 = pd.merge(x_train_task_4,temp,on='card1_card2',how='left')

temp = x_train_task_4.groupby('card1_card2_card_3_card_5')['TransactionAmt'].agg(['mean']).\
rename({'mean':'TransactionAmt_card1_card2_card_3_card_5_mean'},axis=1)
x_train_task_4 = pd.merge(x_train_task_4,temp,on='card1_card2_card_3_card_5',how='left')

temp = x_train_task_4.groupby('card1_card2_card_3_card_5_addr1_addr2')['TransactionAmt'].agg(['mean']).\
rename({'mean':'TransactionAmt_card1_card2_card_3_card_5_addr1_addr2_mean'},axis=1)
x_train_task_4 = pd.merge(x_train_task_4,temp,on='card1_card2_card_3_card_5_addr1_addr2',how='left')

In [124]:
x_train_task_4.head(2)

TransactionDT  TransactionAmt  card1  card2  card3  card5  addr1  addr2   C1   C2   C3   C4   C5   C6   C7   C8   C9  C10  C11  C12   C13  C14     D1  D4    D10    D15  V12  V13  V14  V15  V16  V17  V18  V19  V20  V21  V22  V23  V24  V25  V26  V27  V28  V29  V30  V31  V32  V33  V34  V35  V36  V37  V38  V39  V40  V41  V42  V43  V44  V45  V46  V47  V48  V49  V50  V51  V52  V53  V54  V55  V56  V57  V58  V59  V60  V61  V62  V63  V64  V65  V66  V67  V68  V69  V70  V71  V72  V73  V74  V75  V76  V77  V78  V79  V80  V81  V82  V83  V84  V85  V86  V87  V88  V89  V90  V91  V92  V93  V94  V95  V96  V97  V98  V99  V100  V101  V102  V103  V104  V105  V106  V107  V108  V109  V110  V111  V112  V113  V114  V115  V116  V117  V118  V119  V120  V121  V122  V123  V124  V125  V126   V127   V128  V129   V130   V131  V132  V133  V134  V135  V136  V137  V280  V281  V282  V283  V284  V285  V286  V287  V288  V289  V290  V291  V292  V293  V294  V295  V296  V297  V298  V299  V300  V301  V302  V303  V304  \
0        2916619           218.0   6892  560.0  150.0  226.0  433.0   87.0  3.0  2.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  2.0  0.0  24.0  2.0  220.0 NaN  220.0  220.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  1.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  3.0  2.0  0.0  3.0   2.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   0.0  569.0  452.0   0.0  569.0  452.0   0.0   0.0   0.0   0.0   0.0   0.0   2.0   0.0   0.0   0.0   0.0   3.0   0.0   2.0   0.0   0.0   1.0   1.0   1.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
1        2600138            50.0   2922  583.0  150.0  226.0  299.0   87.0  1.0  1.0  0.0  1.0  0.0  1.0  0.0  1.0  0.0  1.0  1.0  0.0   1.0  1.0    0.0 NaN    NaN    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  0.0  0.0  0.0  0.0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   0.0    0.0    0.0   0.0    0.0    0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0   1.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0   1.0   1.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0   1.0   1.0   

   V305  V306   V307   V308  V309   V310  V311   V312  V313  V314  V315  V316  V317  V318  V319  V320  V321  year  month  week_day  hour  day card1_card2 card1_card2_card_3_card_5 card1_card2_card_3_card_5_addr1_addr2         card1_freq_enc         card2_freq_enc      card3_freq_enc      card4_freq_enc      card5_freq_enc       card6_freq_enc        addr1_freq_enc      addr2_freq_enc card4   card6  TransactionAmt_card1_mean  TransactionAmt_card2_mean  TransactionAmt_card3_mean  TransactionAmt_card5_mean  TransactionAmt_card4_mean  TransactionAmt_card6_mean  TransactionAmt_card1_card2_mean  TransactionAmt_card1_card2_card_3_card_5_mean  TransactionAmt_card1_card2_card_3_card_5_addr1_addr2_mean  
0   1.0   0.0  569.0  452.0   0.0  569.0   0.0  452.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  2017     11         4    18    3  6892 560.0    6892 560.0 150.0 226.0     6892 560.0 150.0 226.0 433.0 87.0  0.0003111111111111111  0.0004359034098168454  0.8791389502066698  0.6582370370370371  0.514

In [125]:
temp = x_test_task_4.groupby('card1')['TransactionAmt'].agg(['mean']).rename({'mean':'TransactionAmt_card1_mean'},axis=1)
x_test_task_4 = pd.merge(x_test_task_4,temp,on='card1',how='left')
temp = x_test_task_4.groupby('card2')['TransactionAmt'].agg(['mean']).rename({'mean':'TransactionAmt_card2_mean'},axis=1)
x_test_task_4 = pd.merge(x_test_task_4,temp,on='card2',how='left')
temp = x_train_task_4.groupby('card3')['TransactionAmt'].agg(['mean']).rename({'mean':'TransactionAmt_card3_mean'},axis=1)
x_test_task_4 = pd.merge(x_test_task_4,temp,on='card3',how='left')
temp = x_test_task_4.groupby('card5')['TransactionAmt'].agg(['mean']).rename({'mean':'TransactionAmt_card5_mean'},axis=1)
x_test_task_4 = pd.merge(x_test_task_4,temp,on='card5',how='left')
temp = x_test_task_4.groupby('card4')['TransactionAmt'].agg(['mean']).rename({'mean':'TransactionAmt_card4_mean'},axis=1)
x_test_task_4 = pd.merge(x_test_task_4,temp,on='card4',how='left')
temp = x_test_task_4.groupby('card6')['TransactionAmt'].agg(['mean']).rename({'mean':'TransactionAmt_card6_mean'},axis=1)
x_test_task_4 = pd.merge(x_test_task_4,temp,on='card6',how='left')

In [126]:
temp = x_test_task_4.groupby('card1_card2')['TransactionAmt'].agg(['mean']).\
rename({'mean':'TransactionAmt_card1_card2_mean'},axis=1)
x_test_task_4 = pd.merge(x_test_task_4,temp,on='card1_card2',how='left')

temp = x_test_task_4.groupby('card1_card2_card_3_card_5')['TransactionAmt'].agg(['mean']).\
rename({'mean':'TransactionAmt_card1_card2_card_3_card_5_mean'},axis=1)
x_test_task_4 = pd.merge(x_test_task_4,temp,on='card1_card2_card_3_card_5',how='left')

temp = x_test_task_4.groupby('card1_card2_card_3_card_5_addr1_addr2')['TransactionAmt'].agg(['mean']).\
rename({'mean':'TransactionAmt_card1_card2_card_3_card_5_addr1_addr2_mean'},axis=1)
x_test_task_4 = pd.merge(x_test_task_4,temp,on='card1_card2_card_3_card_5_addr1_addr2',how='left')

In [127]:
x_test_task_4.head(2)

TransactionDT  TransactionAmt  card1  card2  card3  card5  addr1  addr2   C1   C2   C3   C4   C5   C6   C7   C8   C9  C10  C11  C12   C13  C14     D1     D4    D10    D15  V12  V13  V14  V15  V16  V17  V18  V19  V20  V21  V22  V23  V24  V25  V26  V27  V28  V29  V30  V31  V32  V33  V34  V35  V36  V37  V38  V39  V40  V41  V42  V43  V44  V45  V46  V47  V48  V49  V50  V51  V52  V53  V54  V55  V56  V57  V58  V59  V60  V61  V62  V63  V64  V65  V66  V67  V68  V69  V70  V71  V72  V73  V74  V75  V76  V77  V78  V79  V80  V81  V82  V83  V84  V85  V86  V87  V88  V89  V90  V91  V92  V93  V94  V95  V96  V97  V98  V99  V100  V101  V102  V103  V104  V105  V106  V107  V108  V109  V110  V111  V112  V113  V114  V115  V116  V117  V118  V119  V120  V121  V122  V123  V124  V125  V126   V127  V128  V129   V130  V131  V132  V133  V134  V135  V136  V137  V280  V281  V282  V283  V284  V285  V286  V287  V288  V289  V290  V291  V292  V293  V294  V295  V296  V297  V298  V299  V300  V301  V302  V303  V304  \
0        1712256           171.0  15186  480.0  150.0  224.0  299.0   87.0  1.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  2.0  0.0  1.0  0.0  14.0  1.0  409.0  410.0  410.0  410.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  4.0  1.0  0.0  4.0   1.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   0.0  236.0  59.0   0.0  236.0  59.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0   0.0   0.0   0.0   0.0   5.0   0.0   1.0   0.0   0.0   1.0   2.0   1.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
1         108545            50.0   6019  583.0  150.0  226.0  225.0   87.0  1.0  1.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  1.0  1.0  0.0   1.0  1.0    0.0    NaN    NaN    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  0.0  0.0  0.0  0.0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   0.0    0.0   0.0   0.0    0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0   1.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0   1.0   1.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0   1.0   1.0   

   V305  V306   V307  V308  V309   V310  V311  V312  V313  V314  V315  V316  V317  V318  V319  V320  V321  year  month  week_day  hour  day  card1_card2 card1_card2_card_3_card_5 card1_card2_card_3_card_5_addr1_addr2         card1_freq_enc         card2_freq_enc      card3_freq_enc       card4_freq_enc       card5_freq_enc      card6_freq_enc        addr1_freq_enc      addr2_freq_enc       card4   card6  TransactionAmt_card1_mean  TransactionAmt_card2_mean  TransactionAmt_card3_mean  TransactionAmt_card5_mean  TransactionAmt_card4_mean  TransactionAmt_card6_mean  TransactionAmt_card1_card2_mean  TransactionAmt_card1_card2_card_3_card_5_mean  TransactionAmt_card1_card2_card_3_card_5_addr1_addr2_mean  
0   1.0   0.0  276.0  59.0   0.0  276.0   0.0  59.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  2017     10         4    19   20  15186 480.0   15186 480.0 150.0 224.0    15186 480.0 150.0 224.0 299.0 87.0  0.0002666666666666667  0.0034512316159884507  0.8815307006822374  0.303644444444444

In [128]:
categorical_features = [#'card1_card2',
#  'card1_card2_card_3_card_5',
#  'card1_card2_card_3_card_5_addr1_addr2',
#  'card1_freq_enc',
#  'card2_freq_enc',
#  'card3_freq_enc',
#  'card4_freq_enc',
#  'card5_freq_enc',
#  'card6_freq_enc',
#  'addr1_freq_enc',
#  'addr2_freq_enc',
#  'card4',
#  'card6',
 'TransactionAmt_card1_mean',
 'TransactionAmt_card2_mean',
 'TransactionAmt_card3_mean',                       
 'TransactionAmt_card5_mean',
 'TransactionAmt_card4_mean',
 'TransactionAmt_card6_mean',
 'TransactionAmt_card1_card2_mean',
 'TransactionAmt_card1_card2_card_3_card_5_mean',
 'TransactionAmt_card1_card2_card_3_card_5_addr1_addr2_mean',
]

In [129]:
###x_train_task_4[categorical_features] = x_train_task_4[categorical_features].astype(str)
###x_test_task_4[categorical_features] = x_test_task_4[categorical_features].astype(str)

### CatBoost с категориальными признаками

In [130]:
# eval_sets= [
#     (x_train_task_4[xgb_numerical_features + task_1_fields + categorical_features], y_train),
#     (x_test_task_4[xgb_numerical_features + task_1_fields + categorical_features], y_test)
# ]

In [131]:
# cb_model.fit(
#     x_train_task_4[xgb_numerical_features + task_1_fields + categorical_features],
#     y_train,
#     cat_features = categorical_features,
#     eval_set=eval_sets)

In [132]:
eval_sets= [
    (x_train_task_4[xgb_numerical_features + categorical_features], y_train),
    (x_test_task_4[xgb_numerical_features + categorical_features], y_test)
]

In [133]:
cb_model.fit(
    x_train_task_4[xgb_numerical_features + categorical_features],
    y_train,
    ### cat_features = categorical_features,
    eval_set=eval_sets)

0:	test: 0.6740226	test1: 0.6630859	best: 0.6630859 (0)	total: 205ms	remaining: 3m 24s
10:	test: 0.7893298	test1: 0.7825786	best: 0.7847464 (8)	total: 1.09s	remaining: 1m 37s
20:	test: 0.8253676	test1: 0.8184220	best: 0.8188058 (19)	total: 1.75s	remaining: 1m 21s
30:	test: 0.8499644	test1: 0.8407726	best: 0.8407726 (30)	total: 2.33s	remaining: 1m 12s
40:	test: 0.8547295	test1: 0.8445045	best: 0.8445045 (40)	total: 2.94s	remaining: 1m 8s
50:	test: 0.8558641	test1: 0.8452847	best: 0.8453157 (49)	total: 3.54s	remaining: 1m 5s
60:	test: 0.8595867	test1: 0.8494737	best: 0.8494737 (60)	total: 4.13s	remaining: 1m 3s
70:	test: 0.8622846	test1: 0.8516733	best: 0.8516733 (70)	total: 4.73s	remaining: 1m 1s
80:	test: 0.8643770	test1: 0.8531365	best: 0.8531365 (80)	total: 5.32s	remaining: 1m
90:	test: 0.8663186	test1: 0.8550837	best: 0.8550837 (90)	total: 5.89s	remaining: 58.8s
100:	test: 0.8708580	test1: 0.8598731	best: 0.8598731 (100)	total: 6.48s	remaining: 57.7s
110:	test: 0.8722877	test1: 0.86

Задание 0:
- bestTest = 0.8827161236
- bestIteration = 419

Задание 4:
- bestTest = 0.8842896115
- bestIteration = 442

Вывод:
- Добавление новых признаков (Задание 4) незначительно улучшило качества модели по сравнению с базовым решением.

## Задание 5:<a class="anchor" id="task5"></a>

Создать признаки на основе отношения: <font color='red'>D15</font> к вычисленной статистике. Статистика - среднее значение / стандартное отклонение <font color='red'>D15</font>, сгруппированное по <font color='red'>card1</font> - <font color='red'>card6, addr1, addr2</font>, и по признакам, созданным в задании 2.

In [135]:
x_train_task_5 = []
x_test_task_5 = []
x_train_task_5 = x_train_task_3.copy()
x_test_task_5 = x_test_task_3.copy()

In [136]:
temp = x_train_task_5.groupby('card1')['D15'].agg(['mean']).rename({'mean':'D15_card1_mean'},axis=1)
x_train_task_5 = pd.merge(x_train_task_5,temp,on='card1',how='left')
temp = x_train_task_5.groupby('card2')['D15'].agg(['mean']).rename({'mean':'D15_card2_mean'},axis=1)
x_train_task_5 = pd.merge(x_train_task_5,temp,on='card2',how='left')
temp = x_train_task_5.groupby('card3')['D15'].agg(['mean']).rename({'mean':'D15_card3_mean'},axis=1)
x_train_task_5 = pd.merge(x_train_task_5,temp,on='card3',how='left')
temp = x_train_task_5.groupby('card5')['D15'].agg(['mean']).rename({'mean':'D15_card5_mean'},axis=1)
x_train_task_5 = pd.merge(x_train_task_5,temp,on='card5',how='left')
temp = x_train_task_5.groupby('card4')['D15'].agg(['mean']).rename({'mean':'D15_card4_mean'},axis=1)
x_train_task_5 = pd.merge(x_train_task_5,temp,on='card4',how='left')
temp = x_train_task_5.groupby('card6')['D15'].agg(['mean']).rename({'mean':'D15_card6_mean'},axis=1)
x_train_task_5 = pd.merge(x_train_task_5,temp,on='card6',how='left')

In [137]:
temp = x_train_task_5.groupby('card1_card2')['D15'].agg(['mean']).\
rename({'mean':'D15_card1_card2_mean'},axis=1)
x_train_task_5 = pd.merge(x_train_task_5,temp,on='card1_card2',how='left')

temp = x_train_task_5.groupby('card1_card2_card_3_card_5')['D15'].agg(['mean']).\
rename({'mean':'D15_card1_card2_card_3_card_5_mean'},axis=1)
x_train_task_5 = pd.merge(x_train_task_5,temp,on='card1_card2_card_3_card_5',how='left')

temp = x_train_task_5.groupby('card1_card2_card_3_card_5_addr1_addr2')['D15'].agg(['mean']).\
rename({'mean':'D15_card1_card2_card_3_card_5_addr1_addr2_mean'},axis=1)
x_train_task_5 = pd.merge(x_train_task_5,temp,on='card1_card2_card_3_card_5_addr1_addr2',how='left')

In [138]:
temp = x_test_task_5.groupby('card1')['D15'].agg(['mean']).rename({'mean':'D15_card1_mean'},axis=1)
x_test_task_5 = pd.merge(x_test_task_5,temp,on='card1',how='left')
temp = x_test_task_5.groupby('card2')['D15'].agg(['mean']).rename({'mean':'D15_card2_mean'},axis=1)
x_test_task_5 = pd.merge(x_test_task_5,temp,on='card2',how='left')
temp = x_test_task_5.groupby('card3')['D15'].agg(['mean']).rename({'mean':'D15_card3_mean'},axis=1)
x_test_task_5 = pd.merge(x_test_task_5,temp,on='card3',how='left')
temp = x_test_task_5.groupby('card5')['D15'].agg(['mean']).rename({'mean':'D15_card5_mean'},axis=1)
x_test_task_5 = pd.merge(x_test_task_5,temp,on='card5',how='left')
temp = x_test_task_5.groupby('card4')['D15'].agg(['mean']).rename({'mean':'D15_card4_mean'},axis=1)
x_test_task_5 = pd.merge(x_test_task_5,temp,on='card4',how='left')
temp = x_test_task_5.groupby('card6')['D15'].agg(['mean']).rename({'mean':'D15_card6_mean'},axis=1)
x_test_task_5 = pd.merge(x_test_task_5,temp,on='card6',how='left')

In [139]:
temp = x_test_task_5.groupby('card1_card2')['D15'].agg(['mean']).\
rename({'mean':'D15_card1_card2_mean'},axis=1)
x_test_task_5 = pd.merge(x_test_task_5,temp,on='card1_card2',how='left')

temp = x_test_task_5.groupby('card1_card2_card_3_card_5')['D15'].agg(['mean']).\
rename({'mean':'D15_card1_card2_card_3_card_5_mean'},axis=1)
x_test_task_5 = pd.merge(x_test_task_5,temp,on='card1_card2_card_3_card_5',how='left')

temp = x_test_task_5.groupby('card1_card2_card_3_card_5_addr1_addr2')['D15'].agg(['mean']).\
rename({'mean':'D15_card1_card2_card_3_card_5_addr1_addr2_mean'},axis=1)
x_test_task_5 = pd.merge(x_test_task_5,temp,on='card1_card2_card_3_card_5_addr1_addr2',how='left')

In [140]:
categorical_features = [#'card1_card2',
#  'card1_card2_card_3_card_5',
#  'card1_card2_card_3_card_5_addr1_addr2',
#  'card1_freq_enc',
#  'card2_freq_enc',
#  'card3_freq_enc',
#  'card4_freq_enc',
#  'card5_freq_enc',
#  'card6_freq_enc',
#  'addr1_freq_enc',
#  'addr2_freq_enc',
#  'card4',
#  'card6',
 'D15_card1_mean',
 'D15_card2_mean',
 'D15_card3_mean',                       
 'D15_card5_mean',
 'D15_card4_mean',
 'D15_card6_mean',
 'D15_card1_card2_mean',
 'D15_card1_card2_card_3_card_5_mean',
 'D15_card1_card2_card_3_card_5_addr1_addr2_mean',
]

In [141]:
### x_train_task_5[categorical_features] = x_train_task_5[categorical_features].astype(str)
### x_test_task_5[categorical_features] = x_test_task_5[categorical_features].astype(str)

In [142]:
eval_sets= [
    (x_train_task_5[xgb_numerical_features + categorical_features], y_train),
    (x_test_task_5[xgb_numerical_features + categorical_features], y_test)
]

In [143]:
cb_model.fit(
    x_train_task_5[xgb_numerical_features + categorical_features],
    y_train,
    ### cat_features = categorical_features,
    eval_set=eval_sets)

0:	test: 0.6668196	test1: 0.6364106	best: 0.6364106 (0)	total: 226ms	remaining: 3m 45s
10:	test: 0.7837131	test1: 0.7731993	best: 0.7744707 (8)	total: 1.22s	remaining: 1m 50s
20:	test: 0.8208458	test1: 0.8134966	best: 0.8134966 (20)	total: 1.82s	remaining: 1m 24s
30:	test: 0.8484551	test1: 0.8388704	best: 0.8388704 (30)	total: 2.39s	remaining: 1m 14s
40:	test: 0.8539346	test1: 0.8452090	best: 0.8452090 (40)	total: 2.98s	remaining: 1m 9s
50:	test: 0.8556360	test1: 0.8466824	best: 0.8467786 (49)	total: 3.55s	remaining: 1m 6s
60:	test: 0.8594347	test1: 0.8502323	best: 0.8507420 (58)	total: 4.15s	remaining: 1m 3s
70:	test: 0.8611161	test1: 0.8513599	best: 0.8520064 (65)	total: 4.76s	remaining: 1m 2s
80:	test: 0.8641617	test1: 0.8540539	best: 0.8540607 (78)	total: 5.39s	remaining: 1m 1s
90:	test: 0.8652170	test1: 0.8542878	best: 0.8544052 (81)	total: 5.99s	remaining: 59.8s
100:	test: 0.8681806	test1: 0.8576883	best: 0.8577066 (98)	total: 6.64s	remaining: 59.1s
110:	test: 0.8693815	test1: 0.

Задание 0:
- bestTest = 0.8827161236
- bestIteration = 419

Задание 5:
- bestTest = 0.8832494667
- bestIteration = 463

Вывод:
-  Добавление новых признаков (Задание 5) незначительно улучшило качества модели по сравнению с базовым решением.

## Задание 6:<a class="anchor" id="task6"></a>

Выделить дробную часть и целую часть признака <font color='red'>TransactionAmt</font> в два отдельных признака. После создать отдельных признак - логарифм от <font color='red'>TransactionAmt</font>

In [91]:
import math
# print(5.1 - int(5.1))
# x = math.modf(3.456)
# print(x[0])
# print(x[1])

In [92]:
x_train_task_6 = []
x_test_task_6 = []
x_train_task_6 = x_train_task_3.copy()
x_test_task_6 = x_test_task_3.copy()

In [93]:
import math
print(math.modf(45.8978))

def function(x):
    x = math.modf(x)
    return x[1], x[0]

(0.8977999999999966, 45.0)


In [94]:
# x_train_task_1['new_date'], 
x_train_task_6['TransactionAmr_intager'], x_train_task_6['TransactionAmr_fractional'] = \
zip(*x_train_task_6['TransactionAmt'].map(function))

# x_test_task_1['new_date'], 
x_test_task_6['TransactionAmr_intager'], x_test_task_6['TransactionAmr_fractional'] = \
zip(*x_test_task_6['TransactionAmt'].map(function))

In [95]:
# x_train_task_6['TransactionAmr_log'] = zip(*x_train_task_6['TransactionAmt'].map(function_log))
x_train_task_6['TransactionAmr_log'] = np.log(x_train_task_6['TransactionAmt'])
x_test_task_6['TransactionAmr_log'] = np.log(x_test_task_6['TransactionAmt'])

In [96]:
task6_features = [
 'TransactionAmr_intager',
 'TransactionAmr_fractional',
 'TransactionAmr_log',                       
]

In [150]:
#x_train_task_3["TransactionAmt"]

In [149]:
x_train_task_6[task6_features]

TransactionAmr_intager  TransactionAmr_fractional  TransactionAmr_log
141582                   218.0                   0.000000            5.384495
131503                    50.0                   0.000000            3.912023
173925                    77.0                   0.000000            4.343805
177012                    57.0                   0.950001            4.059581
69958                     44.0                   0.000000            3.784190
...                        ...                        ...                 ...
4848                      25.0                   0.000000            3.218876
14879                     40.0                   0.000000            3.688879
36680                     24.0                   0.000000            3.178054
118456                    63.0                   0.950001            4.158102
5139                      59.0                   0.000000            4.077538

[135000 rows x 3 columns]

In [98]:
# eval_sets= [
#     (x_train_task_6[xgb_numerical_features + categorical_features], y_train),
#     (x_test_task_6[xgb_numerical_features + categorical_features], y_test)
# ]

In [99]:
# cb_model.fit(
#     x_train_task_6[xgb_numerical_features + task6_features],
#     y_train,
#     cat_features = categorical_features,
#     eval_set=eval_sets)

In [100]:
eval_sets= [
    (x_train_task_6[xgb_numerical_features + task6_features], y_train),
    (x_test_task_6[xgb_numerical_features + task6_features], y_test)
]

In [101]:
# cb_model = cb.CatBoostClassifier(**cb_params)
# cb_model.fit(x_train_task_6[xgb_numerical_features + task6_features], y_train, eval_set=eval_sets)

In [102]:
cb_model.fit(
    x_train_task_6[xgb_numerical_features + task6_features],
    y_train,
    #cat_features = categorical_features,
    eval_set=eval_sets)

0:	test: 0.6829308	test1: 0.6767559	best: 0.6767559 (0)	total: 165ms	remaining: 2m 44s
10:	test: 0.7729142	test1: 0.7622869	best: 0.7622965 (9)	total: 1.39s	remaining: 2m 4s
20:	test: 0.8279478	test1: 0.8231270	best: 0.8231270 (20)	total: 2.58s	remaining: 2m
30:	test: 0.8414742	test1: 0.8332268	best: 0.8332268 (30)	total: 3.47s	remaining: 1m 48s
40:	test: 0.8490078	test1: 0.8425562	best: 0.8427865 (39)	total: 4.06s	remaining: 1m 34s
50:	test: 0.8524722	test1: 0.8458381	best: 0.8458381 (50)	total: 4.64s	remaining: 1m 26s
60:	test: 0.8575901	test1: 0.8498734	best: 0.8499933 (57)	total: 5.21s	remaining: 1m 20s
70:	test: 0.8582895	test1: 0.8501271	best: 0.8511148 (64)	total: 5.81s	remaining: 1m 16s
80:	test: 0.8625141	test1: 0.8538566	best: 0.8538566 (80)	total: 6.38s	remaining: 1m 12s
90:	test: 0.8654997	test1: 0.8570493	best: 0.8570493 (90)	total: 6.96s	remaining: 1m 9s
100:	test: 0.8685545	test1: 0.8605576	best: 0.8605576 (100)	total: 7.56s	remaining: 1m 7s
110:	test: 0.8696675	test1: 0

Задание 0 (без обработки):
- bestTest = 0.8827161236
- bestIteration = 419


Задание 6:
- bestTest = 0.8828945346
- bestIteration = 443

Вывод:
- Добавление новых признаков (Задание 6) незначительно улучшило качества модели по сравнению с базовым решением.


## Задание 7 (опция):<a class="anchor" id="task7"></a>

Выполнить предварительную подготовку / очистку признаков <font color='red'>P_emaildomain</font> и <font color='red'>R_emaildomain</font> (что и как делать - остается на ваше усмотрение) и сделать <font color='red'>Frequency Encoding</font> для очищенных признаков.

См. "Урок 4 Предварительная обработка признаков/Категориальные признаки/Второй способ". Файл webinar4_features_part1.ipynb.

In [103]:
x_train_task_7 = []
x_test_task_7 = []
x_train_task_7 = x_train.copy()
x_test_task_7 = x_test.copy()

In [158]:
data = []
data_test = []
data = x_train_task_1.copy()
data_test = x_test_task_1.copy()

In [159]:
x_train_task_7[['P_emaildomain', 'R_emaildomain']]

P_emaildomain  R_emaildomain
141582        Unknown        Unknown
131503      yahoo.com      yahoo.com
173925        Unknown        Unknown
177012        aol.com        Unknown
69958         Unknown        Unknown
...               ...            ...
4848    anonymous.com        Unknown
14879         Unknown  anonymous.com
36680         Unknown        Unknown
118456      gmail.com        Unknown
5139        yahoo.com        Unknown

[135000 rows x 2 columns]

In [160]:
x_test_task_7[['P_emaildomain', 'R_emaildomain']]

P_emaildomain  R_emaildomain
78715   anonymous.com        Unknown
907       comcast.net        Unknown
87782       gmail.com      gmail.com
55343     hotmail.com        Unknown
7372    anonymous.com        Unknown
...               ...            ...
4018        yahoo.com        Unknown
79718       gmail.com  anonymous.com
23131         aol.com        Unknown
99884     hotmail.com    hotmail.com
168530        aol.com      yahoo.com

[45000 rows x 2 columns]

In [161]:
freq_encoder = data["P_emaildomain"].value_counts(normalize=True)
data["P_emaildomain_freq_enc"] = data["P_emaildomain"].map(freq_encoder)
freq_encoder = data["R_emaildomain"].value_counts(normalize=True)
data["R_emaildomain_freq_enc"] = data["R_emaildomain"].map(freq_encoder)

In [162]:
freq_encoder = data_test["P_emaildomain"].value_counts(normalize=True)
data_test["P_emaildomain_freq_enc"] = data_test["P_emaildomain"].map(freq_encoder)
freq_encoder = data_test["R_emaildomain"].value_counts(normalize=True)
data_test["R_emaildomain_freq_enc"] = data_test["R_emaildomain"].map(freq_encoder)

In [163]:
data[["P_emaildomain", "P_emaildomain_freq_enc", "R_emaildomain", "R_emaildomain_freq_enc"]]

P_emaildomain  P_emaildomain_freq_enc  R_emaildomain  R_emaildomain_freq_enc
141582        Unknown                0.158096        Unknown                0.665281
131503      yahoo.com                0.160852      yahoo.com                0.031067
173925        Unknown                0.158096        Unknown                0.665281
177012        aol.com                0.048037        Unknown                0.665281
69958         Unknown                0.158096        Unknown                0.665281
...               ...                     ...            ...                     ...
4848    anonymous.com                0.073985        Unknown                0.665281
14879         Unknown                0.158096  anonymous.com                0.054837
36680         Unknown                0.158096        Unknown                0.665281
118456      gmail.com                0.373281        Unknown                0.665281
5139        yahoo.com                0.160852        Unknown                0.665281

[135000 rows x 4 columns]

In [164]:
categorical_features = [
 'P_emaildomain_freq_enc',
 'R_emaildomain_freq_enc'
]

In [165]:
data[categorical_features] = data[categorical_features].astype(str)
data_test[categorical_features] = data_test[categorical_features].astype(str)

In [166]:
eval_sets= [
    (data[xgb_numerical_features + categorical_features], y_train),
    (data_test[xgb_numerical_features + categorical_features], y_test)
]

In [167]:
cb_model.fit(
    data[xgb_numerical_features + categorical_features],
    y_train,
    #cat_features = categorical_features,
    eval_set=eval_sets)

0:	test: 0.6426443	test1: 0.6397917	best: 0.6397917 (0)	total: 173ms	remaining: 2m 52s
10:	test: 0.7842472	test1: 0.7762249	best: 0.7762864 (9)	total: 1.08s	remaining: 1m 37s
20:	test: 0.8205347	test1: 0.8149997	best: 0.8152477 (19)	total: 1.88s	remaining: 1m 27s
30:	test: 0.8417268	test1: 0.8334874	best: 0.8334874 (30)	total: 2.65s	remaining: 1m 22s
40:	test: 0.8489144	test1: 0.8393514	best: 0.8393800 (38)	total: 3.33s	remaining: 1m 17s
50:	test: 0.8552671	test1: 0.8471204	best: 0.8472737 (48)	total: 3.9s	remaining: 1m 12s
60:	test: 0.8553998	test1: 0.8466190	best: 0.8472958 (53)	total: 4.48s	remaining: 1m 8s
70:	test: 0.8579710	test1: 0.8487815	best: 0.8491857 (68)	total: 5.07s	remaining: 1m 6s
80:	test: 0.8619628	test1: 0.8519043	best: 0.8519043 (80)	total: 5.66s	remaining: 1m 4s
90:	test: 0.8633333	test1: 0.8526949	best: 0.8527723 (89)	total: 6.25s	remaining: 1m 2s
100:	test: 0.8659497	test1: 0.8554849	best: 0.8554849 (100)	total: 6.81s	remaining: 1m
110:	test: 0.8682495	test1: 0.8

__Вывод:__

Задание 0 (без обработки):
- bestTest = 0.8827161236
- bestIteration = 419


Задание 1:
- bestTest = 0.8812417137
- bestIteration = 455

Вывод:
- Добавление новых признаков (Задание 1) не дало улучшения качества модели по сравнению с базовым решением.


Задание 2:
- bestTest = 0.9216976237
- bestIteration = 557

Вывод:
- Добавление новых признаков (Задание 2) значительно улучшило качество модели по сравнению с базовым решением.

Задание 3:
- bestTest = 0.9180509792
- bestIteration = 506

Вывод:
- Добавление новых признаков (Задание 3) значительно улучшило качество модели по сравнению с базовым решением.


Задание 4:
- bestTest = 0.8842896115
- bestIteration = 442

Вывод:
- Добавление новых признаков (Задание 4) незначительно улучшило качества модели по сравнению с базовым решением.


Задание 5:
- bestTest = 0.8832494667
- bestIteration = 463

Вывод:
-  Добавление новых признаков (Задание 5) незначительно улучшило качества модели по сравнению с базовым решением.


Задание 6:
- bestTest = 0.8828945346
- bestIteration = 443

Вывод:
- Добавление новых признаков (Задание 6) незначительно улучшило качества модели по сравнению с базовым решением.


Задание 7:
- bestTest = 0.8859097396
- bestIteration = 458


Вывод:
- Добавление новых признаков (Задание 7) улучшило качества модели по сравнению с базовым решением.
